### PR TITLE
Removes SYNC-3364 [v110] history configuration cleanup

### PR DESCRIPTION
**Pull request by tarikeshaq**
# Disclaimer
⚠️ **This is a very large PR and I'm planning to break it down into smaller ones, but wanted to share out the cleanups I'm working on. If you're curious look at the code, but mostly reviewing the PR description is what I'm looking for.** ⚠️

## Overview

With history using Application services places (expected to rollout remotely in 108) we can start cleaning up a lot of legacy code! This PR is **incomplete** needs a **TON** of testing and is a work in progress. With that preface, here's what's included:

### Removal of configuration
- We added configuration in 108, so we can remotely configure the Places APIs. This PR removes that configuration (assuming places ships OK in 108, we will not land this until the rollout is complete and considered OK)
- This PR removes that configuration completely and enables the places API by default

### Removal of old APIs
- With the configuration gone, we can now safely remove the old SqlHistory APIs related to history storage and sync, this PR does that

### Refactoring of SQLiteHistory
- The SQLiteHistory was used to manage:
  - History
  - TopSites
  - Pinned Sites
  - Favicons
  - History Sync
- After switching to the new history, we can simplify it so that it only manages the features that are not yet available in places:
   - Pinned Sites
   - Favicons
- This PR includes a ton of refactoring to allow that to happen
- **NOTE**: because the refactor touches many surfaces  (including top sites) we should definitely run a round of QA before any of it lands. I'm happy to schedule meetings to walk through the changes too

### Renaming SQLiteHistory to BrowserDBSQLite
- After this PR SQLiteHistory won't manage history. So renamed it to `BrowserDBSQLite` so it's a bit more generic.

### Removing Caching for top sites
- We have a caching layer in SQLiteHistory for top sites, we remove that and call directly into places instead the same way android does

### Deletion of dead code and redundant tests
- There is quite a bit of dead code, and tests. This PR has the removal of that too which accounts to about half the changes here

## Refactors not in this PR
There are a few changes that I intentionally left out of here, because I believe they have a bigger scope or should be done even more carefully. **THIS IS AN INCOMPLETE LIST AND I WILL KEEP ADDING TO IT**

1. Complete removal of user's data in their history tables, from there, there will be no going back
1. Complete removal of the table creation for all the tables we no longer use, it should be safe to remove though.

### CODE REVIEW

**Comment by tarikeshaq** [comment]
Got around to removing the dead tests and refactoring any history tests to test places properly - next step will be to split this apart a little and make sure there is enough context on what's going on here

**Review by tarikeshaq — COMMENTED** [review]
A short self-review as I clean things up more 

If anyone is following along, I want to repeat that I have no intention of landing all of this at once

The idea is to breakdown which of those cleanups are ones we want and land those one by one

File: Client/Frontend/Settings/ClearPrivateDataTableViewController.swift [hunk 1]
```diff
@@ -159,24 +159,13 @@ class ClearPrivateDataTableViewController: ThemedTableViewController {
                 }
             }
 
+            let alert: UIAlertController
             if self.toggles[HistoryClearableIndex] && profile.hasAccount() {
-                profile.syncManager.hasSyncedHistory().uponQueue(.main) { yes in
```
Comment (tarikeshaq) [inline Client/Frontend/Settings/ClearPrivateDataTableViewController.swift:163]: I can bring this back, but this was a query for the purpose of determining if we should display:

"This action will clear all of your private data, including history from your synced devices"

or

"This action will clear all of your private data. It cannot be undone."

I removed it because the query would be complicated in the new history, and having the "including history from your synced devices" seems OK?

File: Storage/SQL/SQLitePinnedSites.swift [hunk 2]
```diff
@@ -0,0 +1,92 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
```
Comment (tarikeshaq) [inline Storage/SQL/SQLitePinnedSites.swift:1]: This file was renamed from SQLiteHistory

**Comment by lmarceau** [comment]
Swiftlint warnings not related to this PR, will be fixed with changes [here](https://github.com/mozilla-mobile/firefox-ios/pull/12512/commits/ecd4a4ad8712485930bee377c8fbdd0c0071eeec)

**Comment by nbhasin2** [comment]
@tarikeshaq have you already done a walkthrough of this PR?
This is too big imo to be reviewed here on github

**Comment by tarikeshaq** [comment]
@nbhasin2 I did a walk-through with @OrlaM and @lmarceau, some notes as well in https://docs.google.com/document/d/1R7Fko3aPc9bNHc3FazSdHJlcPk1LZeX8lbvkzTvDUxA/edit and https://docs.google.com/document/d/11smb_1bXMkHDHA-ZwdsWIoHFeGXgDajO2S4nYmcyrPQ/edit

**Comment by nbhasin2** [comment]
Perfect @tarikeshaq, I will skim through this although you are in good hands of @OrlaM and @lmarceau for the review 🙂
lmk if I can help in any way!!

**Review by OrlaM — CHANGES_REQUESTED** [review]
One small comment, otherwise looks good. 
At some point next week we should create a build off this branch to send to QA for testing.

File: Client/Frontend/Home/TopSites/DataManagement/TopSiteHistoryManager.swift [hunk 3]
```diff
@@ -52,10 +34,7 @@ class TopSiteHistoryManager: DataObserver {
     }
 
     func removeTopSite(site: Site) {
-        profile.history.removeFromPinnedTopSites(site).uponQueue(dataQueue) { [weak self] result in
```
Comment (lmarceau) [inline Client/Frontend/Home/TopSites/DataManagement/TopSiteHistoryManager.swift:55]: Since we removed dataQueue usage here, we can remove that queue entirely I think (not used anymore)

File: Client/Frontend/Library/HistoryPanel/HistoryPanelViewModel.swift [hunk 4]
```diff
@@ -291,55 +278,26 @@ class HistoryPanelViewModel: Loggable, FeatureFlaggable {
...
-                        // as the title
-                        title = info.url
-                    }
-                    let site = Site(url: info.url, title: title)
-                    site.latestVisit = Visit(date: UInt64(info.timestamp) * 1000)
-                    return site
-                }.uniqued()
-                ret.fill(Maybe(success: ArrayCursor(data: sites)))
+            DispatchQueue.main.asyncAfter(deadline: .now() + .milliseconds(100)) {
+                self.isFetchInProgress = false
```
Comment (lmarceau) [inline Client/Frontend/Library/HistoryPanel/HistoryPanelViewModel.swift:295]: Could be worth keeping this comment ? Since we still have the 100ms delay
```
                // Force 100ms delay between resolution of the last batch of results
                // and the next time `fetchData()` can be called.
```

File: Client/Frontend/Library/HistoryPanel/HistoryPanelViewModel.swift [hunk 5]
```diff
@@ -291,55 +278,26 @@ class HistoryPanelViewModel: Loggable, FeatureFlaggable {
...
-                    self.browserLog.debug("currentFetchOffset is: \(self.currentFetchOffset)")
-                }
-                return deferMaybe(result)
+        let deferred = profile.places.getSitesWithBound(
+            limit: queryFetchLimit,
+            offset: currentFetchOffset,
+            excludedTypes: VisitTransitionSet(0)
+        )
+
+        let ret = Deferred<Maybe<Cursor<Site>>>()
```
Comment (lmarceau) [inline Client/Frontend/Library/HistoryPanel/HistoryPanelViewModel.swift:287]: nit; rename to result or something else. I'm not a fan of abbreviation
```suggestion
        let result = Deferred<Maybe<Cursor<Site>>>()
```

File: Client/Utils/HistoryDeletionUtility.swift [hunk 6]
```diff
@@ -102,36 +99,13 @@ class HistoryDeletionUtility: HistoryDeletionProtocol {
...
-                    .uponQueue(.global(qos: .userInteractive)) { result in
-                        completion(result.isSuccess)
-                    }
-            case .new:
-                _ = profile.history
-                    .clearHistory()
-                profile.places.deleteEverythingHistory().uponQueue(.global(qos: .userInteractive)) { result in
-                    completion(result.isSuccess)
-                }
+            profile.places.deleteEverythingHistory().uponQueue(.global(qos: .userInteractive)) { result in
```
Comment (lmarceau) [inline Client/Utils/HistoryDeletionUtility.swift:102]: @OrlaM's question about queue priority applies here too I think. Why .userInteractive and not default?

File: Client/Utils/HistoryDeletionUtility.swift [hunk 7]
```diff
@@ -102,36 +99,13 @@ class HistoryDeletionUtility: HistoryDeletionProtocol {
...
-                        completion(result.isSuccess)
-                    }
-            case .new:
-                // for now we also delete the old history to keep the browser.db up to date
-                _ = profile.history
-                    .removeHistoryFromDate(date)
-                profile.places.deleteVisitsBetween(date).uponQueue(.global(qos: .userInteractive)) { result in
-                    completion(result.isSuccess)
-                }
+            profile.places.deleteVisitsBetween(date).uponQueue(.global(qos: .userInteractive)) { result in
```
Comment (lmarceau) [inline Client/Utils/HistoryDeletionUtility.swift:107]: @OrlaM's question about queue priority applies here too I think. Why .userInteractive and not default?

File: Providers/Profile.swift [hunk 8]
```diff
@@ -438,70 +404,52 @@ open class BrowserProfile: Profile {
...
     lazy var queue: TabQueue = {
-        withExtendedLifetime(self.history) {
+        withExtendedLifetime(self.legacyPlaces) {
             return SQLiteQueue(db: self.database)
         }
     }()
 
     /**
-     * Favicons, history, and tabs are all stored in one intermeshed
+     * Favicons, and pinned sites, and are stored in one intermeshed
```
Comment (lmarceau) [inline Providers/Profile.swift:413]:
```suggestion
     * Favicons and pinned sites are stored in one intermeshed
```

File: Providers/Profile.swift [hunk 9]
```diff
@@ -438,70 +404,52 @@ open class BrowserProfile: Profile {
...
-                errCallback: errCallback)
-        case .real:
-            self.places.migrateHistory(
-                dbPath: self.browserDbPath,
-                lastSyncTimestamp: lastSyncTimestamp,
-                completion: callback,
-                errCallback: errCallback
-            )
+    public func migrateHistoryToPlaces(callback: @escaping (HistoryMigrationResult) -> Void, errCallback: @escaping (Error?) -> Void) {
+        guard FileManager.default.fileExists(atPath: self.browserDbPath) else {
```
Comment (lmarceau) [inline Providers/Profile.swift:440]: nit, `self.` can be removed I think
```suggestion
        guard FileManager.default.fileExists(atPath: browserDbPath) else {
```

File: Providers/Profile.swift [hunk 10]
```diff
@@ -438,70 +404,52 @@ open class BrowserProfile: Profile {
...
+            // This is the user's first run of the app, they don't have a browserDB, so lets report a successful
+            // migration with zero visits
+            callback(HistoryMigrationResult(numTotal: 0, numSucceeded: 0, numFailed: 0, totalDuration: 0))
+            return
         }
+        let lastSyncTimestamp = Int64(self.syncManager.lastSyncFinishTime ?? 0)
+        self.places.migrateHistory(
+            dbPath: self.browserDbPath,
```
Comment (lmarceau) [inline Providers/Profile.swift:448]: nit, self
```suggestion
        let lastSyncTimestamp = Int64(syncManager.lastSyncFinishTime ?? 0)
        places.migrateHistory(
            dbPath:browserDbPath,
```

File: Providers/Profile.swift [hunk 11]
```diff
@@ -1089,12 +1027,6 @@ open class BrowserProfile: Profile {
             }
         }
 
-        fileprivate func syncLegacyHistoryWithDelegate(_ delegate: SyncDelegate, prefs: Prefs, ready: Ready, why: SyncReason) -> SyncResult {
-            log.debug("Syncing history to storage.")
-            let historySynchronizer = ready.synchronizer(HistorySynchronizer.self, delegate: delegate, prefs: prefs, why: why)
-            return historySynchronizer.synchronizeLocalHistory(self.profile.history, withServer: ready.client, info: ready.info, greenLight: self.greenLight())
```
Comment (lmarceau) [inline Providers/Profile.swift:1095]: `func greenLight()` should be removed since not used anymore by removing this code

File: Providers/TopSitesProvider.swift [hunk 12]
```diff
@@ -43,15 +43,26 @@ extension TopSitesProvider {
...
-         prefs: Prefs) {
-        self.browserHistoryFetcher = browserHistoryFetcher
+    init(
+        placesFetcher: RustPlaces,
+        pinnedSiteFetcher: PinnedSites,
+        prefs: Prefs
+    ) {
+        self.placesFetcher = placesFetcher
+        // It's possible that the top sites fetch is the
+        // they very first use of places, lets make sure that
```
Comment (lmarceau) [inline Providers/TopSitesProvider.swift:60]: "Is the they" in the sentence feels like we have an extra word there

File: Shared/Prefs.swift [hunk 13]
```diff
@@ -127,7 +127,7 @@ public struct PrefsKeys {
     public static let WallpaperManagerLogoSwitchPreference = "WallpaperManagerLogoSwitchPreference"
 
     // Application Services Migration to Places DB
-    public static let NewPlacesAPIDefaultKey = "NewPlacesAPI"
+    public static let HistoryMigratedToPlacesKey = "PlacesHistoryMigrationRanreal"
```
Comment (lmarceau) [inline Shared/Prefs.swift:130]: I'm not sure what `Ran real` means ? 
```suggestion
    public static let HistoryMigratedToPlacesKey = "PlacesHistoryMigrationRanReal"
```

File: Storage/PinnedSites.swift [hunk 14]
```diff
@@ -0,0 +1,16 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0
+
+import Shared
+
+/**
+ * A protocol to manage pinned sites in BrowserDB
+ */
```
Comment (lmarceau) [inline Storage/PinnedSites.swift:9]: nit; comment format
```suggestion
/// A protocol to manage pinned sites in BrowserDB
```

File: Storage/Rust/RustPlaces.swift [hunk 15]
```diff
@@ -607,9 +610,55 @@ extension RustPlaces {
...
     }
 
-    public func getTopFrecentSiteInfos(limit: Int, thresholdOption: FrecencyThresholdOption) -> Deferred<Maybe<[TopFrecentSiteInfo]>> {
-        return withReader { connection in
+    public func getTopFrecentSiteInfos(limit: Int, thresholdOption: FrecencyThresholdOption) -> Deferred<Maybe<[Site]>> {
+        let deferred: Deferred<Maybe<[TopFrecentSiteInfo]>> = withReader { connection in
             return try connection.getTopFrecentSiteInfos(numItems: Int32(limit), thresholdOption: thresholdOption)
         }
+
+        let ret = Deferred<Maybe<[Site]>>()
```
Comment (lmarceau) [inline Storage/Rust/RustPlaces.swift:618]: nit, rename to something else than `ret`
```suggestion
        let returnValue = Deferred<Maybe<[Site]>>()
```

File: Storage/SQL/BrowserDB.swift [hunk 16]
```diff
@@ -246,3 +246,21 @@ open class BrowserDB {
           >>== { deferMaybe($0[0] ?? true) }
     }
 }
+
+/**
+ * The sqlite-backed implementation of the history protocol.
+ * Currenlty only supports pinned sites and favicons
+ */
```
Comment (lmarceau) [inline Storage/SQL/BrowserDB.swift:253]: nit; comment format
```suggestion
/// The sqlite-backed implementation of the history protocol.
/// Currenlty only supports pinned sites and favicons
```

File: Tests/ClientTests/Frontend/Browser/TopSitesHelperTests.swift [hunk 17]
```diff
@@ -9,11 +9,35 @@ import Shared
 import XCTest
 
 class TopSitesHelperTests: XCTestCase {
+    let files = MockFiles()
+    fileprivate func deleteDatabases() {
```
Comment (lmarceau) [inline Tests/ClientTests/Frontend/Browser/TopSitesHelperTests.swift:13]:
```suggestion
    private func deleteDatabases() {
```

File: Client/Utils/HistoryDeletionUtility.swift (hunk 6 above)
Comment (tarikeshaq) [reply Client/Utils/HistoryDeletionUtility.swift:97]: heh that's what the old code was like, and I'm not sure if there is good context 😬 Could be a good candidate for a "investigate all the high priority queue requests" follow up though!

File: Client/Utils/HistoryDeletionUtility.swift (hunk 7 above)
Comment (tarikeshaq) [reply Client/Utils/HistoryDeletionUtility.swift:102]: ditto on me not changing this, but maybe I should?

File: Client/Application/AppLaunchUtil.swift [hunk 18]
```diff
@@ -93,6 +66,11 @@ class AppLaunchUtil {
         // has been started.
         initializeExperiments()
 
+        // We migrate history from browser db to places if it hasn't already
+        DispatchQueue.global(qos: .userInitiated).async {
```
[the remark this answers is not in the document]
Comment (tarikeshaq) [reply Client/Application/AppLaunchUtil.swift]: Good point on the priority, it might mean users get their history slightly delayed but from my testing it's still okay be default

File: Shared/Prefs.swift (hunk 13 above)
Comment (tarikeshaq) [reply Shared/Prefs.swift:134]: Yeah the context here is that the `real` at the end used to come from the nimbus configuration - i.e if a user was migrated due to a nimbus rollout (which I'm evaluating based on the experiment) they don't get migrated again when they upgrade to 110

File: Client/Utils/HistoryDeletionUtility.swift (hunk 6 above)
Comment (lmarceau) [reply Client/Utils/HistoryDeletionUtility.swift:97]: Yeah, I might do that related to https://mozilla-hub.atlassian.net/browse/FXIOS-5343

**Comment by lmarceau** [comment]
Special build created to test this PR with Nightly Build [#24394](https://app.bitrise.io/build/a0cbf236-23df-4687-b48f-a674eb1e88a8). This work won’t be merged in `main` until tested. The goal is to make this PR merged for v110.

**Comment by lmarceau** [comment]
The special build was done testing with task https://github.com/mozilla-mobile/firefox-ios/issues/12674, I think this is ready to be merged.

**Review by OrlaM — APPROVED** [review]

**Comment by tarikeshaq** [comment]
Thanks @lmarceau!! Experiment results are in, and I think we should land this. I cover the results in https://docs.google.com/document/d/1fC55c9m66E18ztQwEFhfmesJdSWfes3eSQLcfaq4Kjo/edit

TLDR:
- 5-6% migrations aren't ending, meaning that we are migrating too much data, in the doc I cover why, we have many users with **a lot** of history.
- I'll have a PR before the end of 110 to upgrade to a new version of application-services that limits the migration based on the above findings
- I'll also have another PR before end of 110 for #12680
- Both PRs will be up and ready before the soft freeze, in the week after we're back from the holidays 
- I have a hope that once we use the new history we'll also see some improvement in reported performance issues, but I'm crossing my fingers on this one.

**Comment by OrlaM** [comment]
Build is green on BR 
image
